addons/binding/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/handler/KeContactP20Handler.java

```diff
@@ -0,0 +1,967 @@
+						boolean error = false;
+
+						if (selKey == listenerKey) {
+							try {
+								InetSocketAddress clientAddress = (InetSocketAddress) theChannel
+										.receive(readBuffer);
+								if (clientAddress.getAddress().equals(
+										permittedClientAddress)) {
+									logger.trace(
+											"Received {} on the listener port from {}",
+											new String(readBuffer.array()),
+											clientAddress);
+									numberBytesRead = readBuffer.position();
+								} else {
+									logger.warn(
+											"Received data from '{}' which is not the permitted remote address '{}'",
+											clientAddress,
+											permittedClientAddress);
+									return null;
+								}
+							} catch (Exception e) {
+								error = true;
+							}
+
+						} else {
+
+							try {
+								numberBytesRead = theChannel.read(readBuffer);
+							} catch (NotYetConnectedException e) {
+								logger.warn(
+										"The channel '{}' is not yet connected: {}",
+										theChannel, e.getMessage());
+								error = true;
+							} catch (IOException e) {
+								// If some other I/O error occurs
+								logger.warn(
+										"An IO exception occured on chanel '{}': {}",
+										theChannel, e.getMessage());
+								error = true;
+							}
+						}
+
+						if (numberBytesRead == -1) {
+							error = true;
+						}
+
+						if (error) {
+							logger.info(
+									"Disconnecting '{}' because of a socket error",
+									getThing().getUID().toString());
+							try {
+								theChannel.close();
+							} catch (IOException e) {
+								logger.warn(
+										"An exception occurred while closing the channel '{}': {}",
+										datagramChannel, e.getMessage());
+							}
+
+							onConnectionLost();
+
+						} else {
+							readBuffer.flip();
+							return readBuffer;
+						}
+					}
+				}
+			}
+
+			return null;
+		} finally {
+			lock.unlock();
+		}
+	}
+
+	protected void onWritable(ByteBuffer buffer, DatagramChannel theChannel) {
+		lock.lock();
+		try {
+
+			SelectionKey theSelectionKey = theChannel.keyFor(selector);
+
+			if (theSelectionKey != null) {
+
+				synchronized (selector) {
+					try {
+						selector.selectNow();
+					} catch (IOException e) {
+						logger.error(
+								"An exception occurred while selecting: {}",
+								e.getMessage());
+					}
+				}
+
+				Iterator<SelectionKey> it = selector.selectedKeys().iterator();
+				while (it.hasNext()) {
+					SelectionKey selKey = (SelectionKey) it.next();
+					it.remove();
+					if (selKey.isValid() && selKey.isWritable()
+							&& selKey == theSelectionKey) {
+
+						boolean error = false;
+						buffer.rewind();
+
+						try {
+							logger.trace(
+									"Sending '{}' on the channel '{}'->'{}'",
+									new Object[] { new String(buffer.array()),
+											theChannel.getLocalAddress(),
+											theChannel.getRemoteAddress() });
+							theChannel.write(buffer);
+						} catch (NotYetConnectedException e) {
+							logger.warn(
```

Comment:
again, all these warn messages should not occur in "normal" operation - it should indicate that there is something very wrong.